actors/evm/src/interpreter/instructions/ext.rs

```diff
@@ -48,40 +58,61 @@ pub fn extcodecopy(
+        CodeCid::NotFound | CodeCid::Account => Vec::new(),
+        // calling EXTCODECOPY on native actors results with a single byte 0xFE which solidtiy uses for its `assert`/`throw` methods
+        // and in general invalid EVM bytecode
+        _ => vec![0xFE],
+    };
 
     copy_to_memory(&mut state.memory, dest_offset, size, data_offset, bytecode.as_slice(), true)
 }
 
-pub fn get_evm_bytecode_cid(rt: &impl Runtime, addr: U256) -> Result<Cid, StatusCode> {
-    let addr: EthAddress = addr.into();
-    let addr: Address = addr.try_into()?;
-    // TODO: just return none in most of these cases?
-    let actor_id = rt.resolve_address(&addr).ok_or_else(|| {
-        StatusCode::InvalidArgument("failed to resolve address".to_string())
-        // TODO better error code
-    })?;
-
-    let evm_cid = rt.get_code_cid_for_type(Type::EVM);
-    let target_cid = rt.get_actor_code_cid(&actor_id);
+#[derive(Debug)]
+pub enum CodeCid {
+    /// EVM Address and the CID of the actor (not the bytecode)
+    EVM(Address),
+    Native(Cid),
+    Account,
+    NotFound,
+}
 
-    if Some(evm_cid) != target_cid {
-        return Err(StatusCode::InvalidArgument(
-            "cannot invoke EXTCODESIZE for non-EVM actor".to_string(),
-        )); // TODO better error code
+impl CodeCid {
+    pub fn unwrap_evm<E>(&self, err: E) -> Result<Address, E> {
+        if let CodeCid::EVM(ret) = self {
+            Ok(*ret)
+        } else {
+            Err(err)
+        }
     }
+}
 
-    let cid = rt
-        .send(&addr, crate::Method::GetBytecode as u64, Default::default(), TokenAmount::zero())?
-        .deserialize()?;
-    Ok(cid)
+/// Resolves an address to the address type
+pub fn get_cid_type(rt: &impl Runtime, addr: U256) -> CodeCid {
+    let addr: EthAddress = addr.into();
+
+    addr.try_into()
+        .ok() // into filecoin address
+        .and_then(|addr| rt.resolve_address(&addr)) // resolve actor id
+        .and_then(|id| rt.get_actor_code_cid(&id).map(|cid| (id, cid))) // resolve code cid
+        .map(|(id, cid)| match rt.resolve_builtin_actor_type(&cid) {
+            // TODO part of current account abstraction hack where embryos are accounts
+            Some(Type::Account | Type::Embryo) => CodeCid::Account,
+            Some(Type::EVM) => CodeCid::EVM(Address::new_id(id)),
+            // remaining builtin actors are native
+            _ => CodeCid::Native(cid),
+        })
+        .unwrap_or(CodeCid::NotFound)
 }
 
-pub fn get_evm_bytecode(rt: &impl Runtime, cid: &Cid) -> Result<Vec<u8>, StatusCode> {
+pub fn get_evm_bytecode(rt: &impl Runtime, addr: &Address) -> Result<Vec<u8>, StatusCode> {
+    let cid = rt
+        .send(addr, crate::Method::GetBytecode as u64, Default::default(), TokenAmount::zero())?
+        .deserialize()?;
     let raw_bytecode = rt
         .store()
-        .get(cid) // TODO this is inefficient; should call stat here.
+        .get(&cid) // TODO this is inefficient; should call stat here.
```

Comment:
This is fine. Note: we don't actually _ever_ expect this error as it should be impossible.